engine/index.js

```diff
@@ -0,0 +1,52 @@
+import alexa from './lib/alexa.js';
+import redisHelper from './lib/redis-helper.jd';
+
+function setCurrentSnapshot(redis, snapshotKey) {
+  return new Promise((resolve, reject) => {
+    redis.set('currentSnapshot', snapshotKey, (err) => {
+      if (err) {
+        return reject(err);
+      }
+      resolve();
+    });
+  });
+}
+
+// fetches all required resources from url, builds an Object and returns
+// a promise
+function fetchSite(url) {
+  // XXX write this part
+  throw new Error('not implemented ' + url);
+}
+
+// returns key to the saved snapshot
+function saveSnapshot(redis, topSites, key) {
+  if (!key) {
+    key = new Date().toISOString();
+  }
+  return topSites.forEach((url) => fetchSite(url)
+    .then((siteData) => new Promise((resolve, reject) => {
+      redis.hmset(key, url, siteData, (err) => redis.promiseCallback(err, null, resolve, reject));
+    }))
+  )
+  .then(() => key);
+}
+
+// fetches top sites and saves in database
+// checks if a snapshot is needed
+function fetchAlexaTopSites() {
```

Comment:
Nice! This is a good start. Let's finalize the discussion in #1 before completing the implementation work here.